### PR TITLE
Module #3 - REST API Advanced

### PR DESCRIPTION
**Pull request by vtokarev2006**
This request contains all the tasks from the Rest_API_Advaced topic. Please do a review or let's discuss it.

### CODE REVIEW

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 1]
```diff
@@ -118,7 +132,37 @@ public List<Certificate> getAll(Optional<String> tagName,
...
+        }
+
+        Certificate certificate = em.find(Certificate.class, id);
+        if (certificate == null)
+            throw new ResourceDoesNotExistException("Certificate doesnt exist, id = ");
+
+        if (duration == null && price == null)
+            throw new BadRequestException("Nothing to update for certificate id = " + id);
+
+
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:157]: two new lines

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 2]
```diff
@@ -118,7 +132,37 @@ public List<Certificate> getAll(Optional<String> tagName,
...
+        } catch (NumberFormatException e){
+            throw new BadRequestException("Certificate patch request is malformed, id = " + id);
+        }
+
+        Certificate certificate = em.find(Certificate.class, id);
+        if (certificate == null)
+            throw new ResourceDoesNotExistException("Certificate doesnt exist, id = ");
+
+        if (duration == null && price == null)
+            throw new BadRequestException("Nothing to update for certificate id = " + id);
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:155]: please wrap this with {}

File: src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java [hunk 3]
```diff
@@ -17,8 +20,8 @@
 import java.util.Optional;
 
 @Repository
-@RequiredArgsConstructor
-public class TagRepositoryImpl implements TagRepository {
+@Qualifier("TagRepositoryJdbcTemplate")
+public class TagRepositoryJdbcTemplate implements TagRepository {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java:24]: It has name which contains JdbcTemplate but its a Repository? Why you using Qualifier with exactly same name? This is not needed. Also please name it TagRepositoryImpl

File: src/main/java/com/epam/esm/repository/jpa/TagRepositoryJpa.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @Override
+    public boolean delete(long id) {
+        return false;
+    }
+
+    @Override
+    public Tag findByName(String name) {
+        return null;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jpa/TagRepositoryJpa.java:53]: Null?

File: src/main/java/com/epam/esm/repository/springdata/CertificateCustomRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import jakarta.persistence.TypedQuery;
+import jakarta.persistence.criteria.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+import java.util.*;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/springdata/CertificateCustomRepositoryImpl.java:18]: *

File: src/main/java/com/epam/esm/repository/springdata/OrderSpecification.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Predicate;
+import jakarta.persistence.criteria.Root;
+import org.springframework.data.jpa.domain.Specification;
+
+public class OrderSpecification implements Specification<Order> {
+    @Override
+    public Predicate toPredicate(Root<Order> root, CriteriaQuery<?> query, CriteriaBuilder criteriaBuilder) {
+        return null;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/springdata/OrderSpecification.java:13]: null?

File: src/main/java/com/epam/esm/repository/springdata/TagRepository.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+import java.util.List;
+
+public interface TagRepository extends JpaRepository<Tag, Long> {
+        @Query("select t, count(o) as freq from Order o join User u on(o.user = u) join Certificate c on (o.certificate = c) join c.tags t where u.id = :userId group by t order by freq desc")
+
+/*
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/springdata/TagRepository.java:14]: Why this comment here?

File: src/main/java/com/epam/esm/services/OrderService.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+@Service
+public class OrderService {
+    private final OrderRepository orderRepository;
+    private final com.epam.esm.repository.springdata.OrderRepository orderRepositorySpringData;
+    private final UserRepository userRepository;
+    private final CertificateRepository certificateRepository;
+
+    @Autowired
+    public OrderService(OrderRepository orderRepository, com.epam.esm.repository.springdata.OrderRepository orderRepositorySpringData, UserRepository userRepository, CertificateRepository certificateRepository) {
+        this.orderRepository = orderRepository;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/services/OrderService.java:31]: Why here two order repositories?

File: src/main/java/com/epam/esm/services/TagService.java [hunk 9]
```diff
@@ -1,34 +1,116 @@
...
 @Service
 public class TagService {
-    final private TagRepository tagRepository;
 
-    public TagService(TagRepository tagRepository) {
-        this.tagRepository = tagRepository;
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+
+    //    final private TagRepository tagRepositoryJdbcTemplate;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/services/TagService.java:36]: Remove comments in all code please, if your code is good and clear no comments needed, only if you need a strange decision which needs to be described

File: src/test/java/com/epam/esm/EsmApplicationTests.java [hunk 10]
```diff
@@ -1,15 +1,159 @@
...
+import java.util.stream.IntStream;
 
 @SpringBootTest
 class EsmApplicationTests {
 
+    private final OrderRepository orderRepository;
+
+    private final TagRepository tagRepository;
+
+
```
Comment (Occazion) [inline src/test/java/com/epam/esm/EsmApplicationTests.java:33]: A lots of spaces

**Comment by vtokarev2006** [comment]
1) project requirements added
2) README.md added
3) api/v1/certificates - removed from the project

4) Required for the project repositories are located in the package com.epam.esm.repository.springdata. All other DB access layer classes and interfaces were created during the educational process and stand for educational purposes only. All these old(educational) DB access layer beans are located in dev Spring profile. Services or other parts of the application have no dependencies on these educational classes.

5) All fixes according to review.

File: src/main/java/com/epam/esm/controllers/TagController.java [hunk 11]
```diff
@@ -1,59 +1,91 @@
...
     @DeleteMapping("/{id}")
     public ResponseEntity<String> delete(@PathVariable long id) {
         if (!tagService.delete(id)) {
             throw new ResourceDoesNotExistException("Tag not found, tagId=" + id);
         }
-        return new ResponseEntity<>(HttpStatus.OK);
+        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
+    }
+
+    @GetMapping("/mostWidelyUsedTagOfUserWithHighestCostOfOrders")
```
Comment (Occazion) [inline src/main/java/com/epam/esm/controllers/TagController.java:86]: very ugly url, let's use /summary/{userId} and fetch by user

File: src/main/java/com/epam/esm/domain/Tag.java [hunk 12]
```diff
@@ -1,11 +1,31 @@
...
 
 @Data
 @Builder
-public class Tag {
+@RequiredArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "tags")
+@NamedQuery(query = """
+        select new com.epam.esm.domain.dto.TagOrdersPriceDto(t.id, t.name, sum(o.price))
```
[the remark this answers is not in the document]
Comment (Occazion) [reply src/main/java/com/epam/esm/domain/Tag.java:23]: Don't forget to update this

File: src/main/java/com/epam/esm/repository/CertificateRepository.java [hunk 13]
```diff
@@ -1,18 +1,19 @@
...
-                                    Optional<String> name,
-                                    Optional<String> description,
-                                    Optional<String> orderBy,
-                                    String orderDirection);
+    List<Certificate> fetchCertificatesBySearchParams(Optional<String> tagName,
+                                                      Optional<String> name,
+                                                      Optional<String> description,
+                                                      Optional<String> orderBy,
+                                                      String orderDirection);
+    Certificate partialUpdate(long id, Map<String, String> fields);
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/CertificateRepository.java:17]: Why this called partial update? why just name it updateById and pass certificate with new data?

Comment (Occazion) [reply src/main/java/com/epam/esm/repository/CertificateRepository.java:17]: It's very strange method I didn't see something similar in my work. Commonly we just passing the exact entity(Certificate)

Comment (Occazion) [reply src/main/java/com/epam/esm/repository/CertificateRepository.java:17]: Even if this deprecated please remove this bad choice

File: src/main/java/com/epam/esm/repository/GenericRepository.java [hunk 14]
```diff
@@ -1,12 +1,19 @@
...
+
 import java.util.List;
 import java.util.Optional;
 
+@Deprecated
 interface GenericRepository<T> {
-    Optional<T> get(long id);
-    List<T> getAll();
+    Optional<T> findById(long id);
+
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/GenericRepository.java:11]: remove these spaces, they not needed in 20 lines class

File: src/main/java/com/epam/esm/repository/OrderRepository.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.domain.Order;
+import org.springframework.context.annotation.Profile;
+
+import java.util.List;
+@Deprecated
+public interface OrderRepository extends GenericRepository<Order> {
+    List<Order> getByUserId(long userId);
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/OrderRepository.java:9]: fetchUserById

File: src/main/java/com/epam/esm/repository/TagRepository.java [hunk 16]
```diff
@@ -1,9 +1,13 @@
...
 
 import com.epam.esm.domain.Tag;
+import com.epam.esm.domain.dto.TagOrdersPriceDto;
 
 import java.util.List;
 
+@Deprecated
 public interface TagRepository extends GenericRepository<Tag> {
     Tag findByName(String name);
+    List<TagOrdersPriceDto> getTagSumOrdersPrice(long userId);
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/TagRepository.java:11]: Same point here, fetch

Comment (Occazion) [reply src/main/java/com/epam/esm/repository/TagRepository.java:11]: Use get word only in services

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 17]
```diff
@@ -21,13 +24,23 @@
...
+import java.util.StringJoiner;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
+@Deprecated
 @Repository
-@RequiredArgsConstructor
-public class CertificateRepositoryImpl implements com.epam.esm.repository.CertificateRepository {
+@Profile("dev")
+public class CertificateRepositoryJdbcTemplate implements com.epam.esm.repository.CertificateRepository {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:43]: Why you've using full path to repository class?

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 18]
```diff
@@ -47,35 +60,23 @@ public class CertificateRepositoryImpl implements com.epam.esm.repository.Certif
...
+        this.jdbcTemplate = jdbcTemplate;
+        this.tagRepository = tagRepository;
+        this.namedParameterJdbcTemplate = namedParameterJdbcTemplate;
+        this.em = em;
+        this.dataSource = dataSource;
     }
 
     @Override
-    public Optional<Certificate> get(long id) {
+    public Optional<Certificate> findById(long id) {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:79]: fetchById

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 19]
```diff
@@ -118,16 +113,45 @@ public List<Certificate> getAll(Optional<String> tagName,
         sqlStr += whereStr;
 
         sqlStr += orderBy.isEmpty() ? " ORDER BY c.id" : " ORDER BY " + orderBy.get() + " " + orderDirection + ", c.id";
-        return  namedParameterJdbcTemplate.query(sqlStr, params, new CertificateWithTagsResultSetExtractor());
+        return namedParameterJdbcTemplate.query(sqlStr, params, new CertificateWithTagsResultSetExtractor());
     }
 
+    @Transactional
+    @Override
+    public Certificate partialUpdate(long id, Map<String, String> fields) {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:121]: Really what a strange behaviour, you've named it partialUpdate, why you don't have fullUpdate? Anyways please don't pass Map, just pass Certificate , after that you can copy properties more easily

Comment (Occazion) [reply src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:121]: Find way to copy and update only needed properties, but not with Map

File: src/main/java/com/epam/esm/domain/Tag.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vtokarev2006) [reply src/main/java/com/epam/esm/domain/Tag.java:23]: The named query "Tag_getTagSumOrdersPrice" doesn't use anymore. it's here only for history purposes.

File: src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java [hunk 20]
```diff
@@ -30,13 +33,18 @@ public class TagRepositoryImpl implements TagRepository {
...
+    }
+
     @Override
     public Optional<Tag> get(long id) {
         return Optional.ofNullable(jdbcTemplate.queryForObject(SQL_GET_BY_ID, new TagRowMapper(), id));
     }
 
     @Override
-    public List<Tag> getAll() {
+    public List<Tag> getAll(Pageable pageable) {
```
[the remark this answers is not in the document]
Comment (vtokarev2006) [reply src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java]: It was renamed to findById.

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java (hunk 19 above)
Comment (vtokarev2006) [reply src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:121]: the method was refactored to "Certificate updateById(long id, Certificate certificate)"

File: src/main/java/com/epam/esm/domain/User.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.CreatedDate;
+import org.springframework.data.annotation.LastModifiedDate;
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+import java.io.Serializable;
+import java.time.Instant;
+
+@Data
```
Comment (Occazion) [inline src/main/java/com/epam/esm/domain/User.java:23]: Hm, so many annotation, maybe it's better to create Entity class with all needed annotations?

File: src/main/java/com/epam/esm/exceptions/RestResponseEntityExceptionHandler.java [hunk 22]
```diff
@@ -10,20 +10,17 @@
 public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
 
     @ExceptionHandler(value = TagDuplicateNameException.class)
-    public ResponseEntity<String> handleTagDuplicateNameException(TagDuplicateNameException e) {
-            return new ResponseEntity<>(e.getMessage(), HttpStatus.CONFLICT);
+    public ResponseEntity<ErrorMessage> handleTagDuplicateNameException(TagDuplicateNameException e) {
+        return new ResponseEntity<>(ErrorMessage.builder().msg(e.getMessage()).errorCode(e.getErrorCode()).build(), HttpStatus.CONFLICT);
```
Comment (Occazion) [inline src/main/java/com/epam/esm/exceptions/RestResponseEntityExceptionHandler.java:14]: If this thing needs to be updated with new errors code system?

Comment (Occazion) [inline src/main/java/com/epam/esm/exceptions/RestResponseEntityExceptionHandler.java:14]: Hm, can you use simple constructor?

File: src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java [hunk 23]
```diff
@@ -46,36 +61,24 @@ public class CertificateRepositoryImpl implements com.epam.esm.repository.Certif
...
-        return null;
+    private static final String SQL_DELETE_BY_ID = "DELETE FROM certificates WHERE id = ?";
+    private final JdbcTemplate jdbcTemplate;
+    private final TagRepository tagRepository;
+    private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+    private final EntityManager em;
+    private final DataSource dataSource;
+
+    @Autowired
+    public CertificateRepositoryJdbcTemplate(JdbcTemplate jdbcTemplate, @Qualifier("TagRepositoryJdbcTemplate") TagRepository tagRepository, NamedParameterJdbcTemplate namedParameterJdbcTemplate, EntityManager em, DataSource dataSource) {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/CertificateRepositoryJdbcTemplate.java:72]: This pretty long line, please split up constructor params

File: src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java [hunk 24]
```diff
@@ -67,8 +70,12 @@ public boolean delete(long id) {
...
 
     @Override
-    public Tag findByName(String name) {
+    public Tag fetchByName(String name) {
         return jdbcTemplate.queryForObject(SQL_GET_BY_NAME, new TagRowMapper(), name);
     }
 
-}
+    @Override
+    public List<TagOrdersPriceDto> fetchTagSumOrdersPrice(long userId) {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/jdbctemplate/TagRepositoryJdbcTemplate.java:78]: same naming here, fetchTagSummary

File: src/main/java/com/epam/esm/repository/springdata/CertificateCustomRepositoryImpl.java [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+     *
+     * @param name        - search certificates with name equal the value of the param
+     * @param description - search certificates with description contain the value of the param
+     * @param tagNames    - search certificates with tags name (and condition)
+     * @param pageable    - object determinate pageable behavior
+     * @return - page of Certificates meet params above
+     */
+
+    @Override
+    public Page<Certificate> findCertificatesByNameDescriptionTagNames(Optional<String> name, Optional<String> description, Set<String> tagNames, Pageable pageable) {
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/springdata/CertificateCustomRepositoryImpl.java:56]: Very big method, please split up logic to separate methods

File: src/main/java/com/epam/esm/services/UserService.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+    private final com.epam.esm.repository.springdata.UserRepository userRepository;
```
Comment (Occazion) [inline src/main/java/com/epam/esm/services/UserService.java:20]: Why here full path to class?

Comment (Occazion) [reply src/main/java/com/epam/esm/services/UserService.java:20]: but in imports you have import com.epam.esm.repository.springdata.UserRepository;

File: src/test/java/com/epam/esm/EsmApplicationTests.java [hunk 27]
```diff
@@ -1,15 +1,102 @@
...
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
 @SpringBootTest
 class EsmApplicationTests {
+    private final TagRepository tagRepository;
+    private final Faker faker;
+    private final com.epam.esm.repository.springdata.UserRepository userRepository;
```
Comment (Occazion) [inline src/test/java/com/epam/esm/EsmApplicationTests.java:26]: Why you've using full paths in fields?

File: src/test/java/com/epam/esm/EsmApplicationTests.java [hunk 28]
```diff
@@ -1,15 +1,102 @@
...
+import com.epam.esm.repository.springdata.TagRepository;
+import com.epam.esm.repository.springdata.UserRepository;
+import com.github.javafaker.Faker;
+import jakarta.persistence.EntityManager;
 import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
 
+import java.time.Instant;
+import java.util.*;
```
Comment (Occazion) [inline src/test/java/com/epam/esm/EsmApplicationTests.java:18]: Please do not use *, you can set up it in settings of IDEA

File: src/main/java/com/epam/esm/domain/User.java (hunk 21 above)
Comment (vtokarev2006) [reply src/main/java/com/epam/esm/domain/User.java:23]: BaseEntity class was created. id, createDate, lastUpdateDate fields were moved to the superclass. However, it's not possible to move annotations to the superclass that way.

File: src/main/java/com/epam/esm/repository/OrderRepository.java [hunk 29]
```diff
@@ -0,0 +1,21 @@
+package com.epam.esm.repository;
+
+import com.epam.esm.domain.Order;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface OrderRepository extends JpaRepository<Order, Long> {
+    Page<Order> findByUserId(long userId, Pageable pageable);
+/*
```
Comment (Occazion) [inline src/main/java/com/epam/esm/repository/OrderRepository.java:10]: Is this comment needed?

**Review by Occazion — APPROVED** [review]

**Pull request merged**